django/thunderstore/repository/api/experimental/serializers.py

```diff
@@ -95,3 +96,132 @@ class Meta:
             "has_nsfw_content",
             "categories",
         )
+
+
+class PackageUploadAuthorNameField(serializers.SlugRelatedField):
+    def __init__(self, *args, **kwargs):
+        kwargs["slug_field"] = "name"
+        super().__init__(*args, **kwargs)
+
+    def get_queryset(self):
+        return UploaderIdentity.objects.filter(
+            members__user=self.context["request"].user,
+        )
+
+
+class PackageUploadCategoriesField(serializers.RelatedField):
+    def get_queryset(self):
+        return PackageCategory.objects.filter(
+            community=self.context["request"].community,
+        )
+
+    def to_representation(self, value):
+        return [c.pk for c in value]
+
+    def to_internal_value(self, data):
+        if not isinstance(data, list):
+            raise serializers.ValidationError("Not a list")
+
+        out = []
+        for category_pk in data:
+            category = PackageCategory.objects.filter(pk=category_pk).first()
+            if not category:
+                raise serializers.ValidationError(f"{category_pk} category not found")
+            out.append(category)
+        return out
+
+
+class PackageUploadSerializer(serializers.Serializer):
+    author_name = PackageUploadAuthorNameField()
+    categories = PackageUploadCategoriesField()
+    has_nsfw_content = serializers.BooleanField()
+    pk = serializers.IntegerField(read_only=True)
+    file = serializers.FileField(write_only=True)
+
+    def _create_form(self, data) -> PackageUploadForm:
+        request = self.context["request"]
+        return PackageUploadForm(
+            request.user,
+            data.get("author_name"),
+            request.community,
+            data={
+                "categories": data.get("categories"),
+                "has_nsfw_content": data.get("has_nsfw_content"),
+            },
+            files={"file": data.get("file")},
+        )
+
+    def validate(self, data):
+        form = self._create_form(data)
+        if not form.is_valid():
+            raise serializers.ValidationError("Invalid")
+        return data
+
+    def create(self, validated_data) -> PackageVersion:
+        form = self._create_form(validated_data)
+        form.is_valid()
+        return form.save()
+
+
+class UploaderIdentitySerializer(serializers.ModelSerializer):
+    class Meta:
+        model = UploaderIdentity
+        fields = (
+            "name",
+            "pk",
+        )
+
+    def create(self, *args, **kwargs):
+        raise Exception()
+
+    def update(self, *args, **kwargs):
+        raise Exception()
+
+
+class PackageSerializer(serializers.ModelSerializer):
```

Comment:
I didn't realise there were already some existing similar serializers defined. Maybe we could remove the existing ones and switch to these?

Reply:
Each API should have their own serializers even if it's duplicate code, as to not create dependencies between the API versions. They're intentionally decoupled

Reply:
This is still decoupled as they're in experimental.serializers and don't call other versions, the only difference I can see is the name but I'm not sure why you would need the version in the class name

Reply:
Ah I see there's a duplicate implementation in the same file, I assumed you meant the other package serializers.

The naming isn't really required but I like to keep class names project-scope unique if I can help it, sometimes it can cause annoying collisions (e.g. autoimporting works far better with unique names)